internal/source/image.go

```diff
@@ -0,0 +1,198 @@
+package source
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"strings"
+
+	"github.com/nlepage/go-tarfs"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+	"github.com/operator-framework/rukpak/internal/util"
+)
+
+type Image struct {
+	Client          client.Client
+	KubeClient      kubernetes.Interface
+	ProvisionerName string
+	PodNamespace    string
+	UnpackImage     string
+}
+
+const imageBundleUnpackContainerName = "bundle"
+
+func (i *Image) SourceContent(ctx context.Context, bundle *rukpakv1alpha1.Bundle) (*Result, error) {
+	if bundle.Spec.Source.Type != rukpakv1alpha1.BundleSourceTypeImage {
+		return nil, fmt.Errorf("bundle source type %q not supported", bundle.Spec.Source.Type)
+	}
+	if bundle.Spec.Source.Image == nil {
+		return nil, fmt.Errorf("bundle source image configuration is unset")
+	}
+
+	pod := &corev1.Pod{}
+	op, err := i.ensureUnpackPod(ctx, bundle, pod)
+	if err != nil {
+		return nil, err
+	}
+	if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated || pod.DeletionTimestamp != nil {
+		return &Result{State: StatePending}, nil
+	}
+
+	switch phase := pod.Status.Phase; phase {
+	case corev1.PodPending:
+		return pendingImagePodResult(pod), nil
+	case corev1.PodRunning:
+		return &Result{State: StateUnpacking}, nil
+	case corev1.PodFailed:
+		return nil, i.failedPodResult(ctx, pod)
+	case corev1.PodSucceeded:
+		return i.succeededPodResult(ctx, pod)
+	default:
+		return nil, i.handleUnexpectedPod(ctx, pod)
+	}
+}
+
+func (i *Image) RegisterWatches(c controller.Controller) error {
+	return c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
```

Comment:
Do we need to watch for changes to the object/metadata ConfigMaps(s) for either the git or image based Source implementations?

Reply:
No. The `ConfigMap` watch is the domain of the bundle controller because it is the thing creating/updating config maps after the `Source` has unpacked the content and handed it to the controller in the from of an `fs.FS`.